### PR TITLE
refactor output storage to also store outputs as files

### PR DESCRIPTION
**Pull request by mrijken**
Implements #8

### CODE REVIEW

File: dvb/datascience/pipeline.py [hunk 1]
```diff
@@ -21,6 +24,63 @@ class Status(Enum):
...
+        """
+
+class TransformOutputsMemory(TransformOutputsBase):
+
+    # outputs = None
+
+    def __init__(self):
+        self.outputs: Dict[Tuple[str, int], Dict[str, Any]] = {}
+
+    def save_pipe_output(self, pipe_name: str, transform_nr: int, output: Dict[str, Any]) ->None:
```
Comment (wpbs) [inline dvb/datascience/pipeline.py]:
```suggestion
    def save_pipe_output(self, pipe_name: str, transform_nr: int, output: Dict[str, Any]) -> None:
```
(space after the arrow)

File: dvb/datascience/pipeline.py [hunk 2]
```diff
@@ -21,6 +24,63 @@ class Status(Enum):
...
+    def load_pipe_output(self, pipe_name: str, transform_nr: int) -> Dict[str, Any]:
+        return self.outputs[(pipe_name, transform_nr)]
+
+    def contains_pipe(self, pipe_name: str, transform_nr: int) -> bool:
+        return (pipe_name, transform_nr) in self.outputs
+
+class TransformOutputsFile(TransformOutputsBase):
+
+    def __init__(self, directory='transform_output', file_pattern='{pipe_name}_{transform_nr}.pkl'):
+        self.directory = pathlib.Path(directory)
```
Comment (wpbs) [inline dvb/datascience/pipeline.py]: do you always want to store these files in the same directory as the current one?

File: dvb/datascience/pipeline.py [hunk 3]
```diff
@@ -21,6 +24,63 @@ class Status(Enum):
...
+        self.file_pattern = file_pattern
+
+        self.directory.mkdir(parents=True, exist_ok=True)
+
+    def save_pipe_output(self, pipe_name: str, transform_nr: int,  output: Dict[str, Any]) ->None:
```
Comment (wpbs) [inline dvb/datascience/pipeline.py]:
```suggestion
    def save_pipe_output(self, pipe_name: str, transform_nr: int,  output: Dict[str, Any]) -> None:
```
(space after the arrow)

File: dvb/datascience/pipeline.py [hunk 4]
```diff
@@ -21,6 +24,63 @@ class Status(Enum):
...
+        filename = self.file_pattern.format(pipe_name=pipe_name, transform_nr= transform_nr)
+        with open(self.directory / filename, 'wb') as f:
+            pickle.dump(output, f)
+
+    def load_pipe_output(self, pipe_name: str, transform_nr: int) -> Dict[str, Any]:
+        filename = self.file_pattern.format(pipe_name=pipe_name, transform_nr= transform_nr)
```
Comment (wpbs) [inline dvb/datascience/pipeline.py]:
```suggestion
        filename = self.file_pattern.format(pipe_name=pipe_name, transform_nr=transform_nr)
```
(removed space before transform_nr)

File: dvb/datascience/pipeline.py [hunk 5]
```diff
@@ -64,7 +132,10 @@ def __init__(self):
     def reset_fit(self):
         self.current_transform_nr = -1
         self.transform_status = {}
-        self.transform_outputs = defaultdict(lambda: defaultdict(dict))
+        if self.store_output == 'memory':
```
Comment (wpbs) [inline dvb/datascience/pipeline.py]: Perhaps it is better to let each output transformer to register itself and here just check whether the provided one is contained in the list of available ones? The you don't have to add a line here each time you add a new one (json output, store in database, ...)

File: test/tests/pipeline_test.py [hunk 6]
```diff
@@ -60,6 +57,27 @@ def testPipeline(self):
         )
         self.assertEqual(len(p.get_pipe_output("split")["df"]), 45)
 
+    def testPipelineOutputInFile(self):
```
Comment (wpbs) [inline test/tests/pipeline_test.py]:
```suggestion
    def testPipelineOutputToFile(self):
```

File: dvb/datascience/pipeline.py (hunk 2 above)
Comment (mrijken) [reply dvb/datascience/pipeline.py]: No, it's up to the caller to change the direcory.

File: dvb/datascience/pipeline.py (hunk 5 above)
Comment (mrijken) [reply dvb/datascience/pipeline.py]: I will change it; it will accept  TransformOutputsBase instances as argument.

**Pull request merged**